sql: allow mismatch type numbers in `PREPARE` statement

ZhouXing19 opened the pull request:
Previously, we only allow having the same number of parameters and placeholders
in a `PREPARE` statement. This is not compatible with Postgres14's behavior.

This commit is to loosen the restriction and enable this compatibility.
We now take `max(#placeholders, #parameters)` as the true length
 of parameters of the prepare statement. For each parameter, we first
look at the type deduced from the query stmt. If we can't deduce it, 
we take the type hint for this param.

I.e. we now allow queries such as 

```
PREPARE args_test_many(int, int) as select $1
// 2 parameters, but only 1 placeholder in the query.

PREPARE args_test_few(int) as select $1, $2::int
// 1 parameter, but 2 placeholders in the query.
```

fixes #86375

Release justification: Low risk, high benefit changes to existing functionality
Release note: allow mismatch type numbers in `PREPARE` statement

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86904)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[`pkg/ccl/logictestccl/tests/3node-tenant/generated_test.go` line 1033 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAPvCrS3eeagTfbhx5H:-NAPvCrS3eeagTfbhx5I:b-cshv58) ([raw file](https://github.com/cockroachdb/cockroach/blob/13e9135cb837ed67ab45a6bd6dbc1ed7ccf99067/pkg/ccl/logictestccl/tests/3node-tenant/generated_test.go#L1033)):*
> ```Go
> }
> 
> func TestTenantLogic_jane_test(
> ```

looks like an accidental diff. check out this for how you can add a tmp test https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/154206209/Goland+Tips+and+Tricks#Run-GoLand-debugger-with-logic-tests?
___
*[`pkg/sql/plan_opt.go` line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAPyKya3Jdm7jAd2TY9:-NAPyKyc2GCTF3hlDi_-:b-wlwkyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/13e9135cb837ed67ab45a6bd6dbc1ed7ccf99067/pkg/sql/plan_opt.go#L191)):*
> ```Go
> 
> 	// Fill blank placeholder types with the type hints.
> 	p.semaCtx.Placeholders.MaybeExtendTypes()
> ```

why is this the right place to call it? can the MaybeExtendTypes be done as part of `placeholders.Init()`?

another reason i ask is that this part of the code is shared between the pgwire-level `Parse` logic and the sql-level PREPARE logic, so i just want to make sure it's ok to do this for both.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/ccl/logictestccl/tests/3node-tenant/generated_test.go` line 1033 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAPvCrS3eeagTfbhx5H:-NAQSYnP0X44OimrmoFc:b-r5dyfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/13e9135cb837ed67ab45a6bd6dbc1ed7ccf99067/pkg/ccl/logictestccl/tests/3node-tenant/generated_test.go#L1033)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

looks like an accidental diff. check out this for how you can add a tmp test https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/154206209/Goland+Tips+and+Tricks#Run-GoLand-debugger-with-logic-tests?
</blockquote></details>

Yeah, it was a mistake :sweat_smile: Deleted them!
___
*[`pkg/sql/plan_opt.go` line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAPyKya3Jdm7jAd2TY9:-NAQSipm0to9tNJwjPBT:b40h9n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/13e9135cb837ed67ab45a6bd6dbc1ed7ccf99067/pkg/sql/plan_opt.go#L191)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why is this the right place to call it? can the MaybeExtendTypes be done as part of `placeholders.Init()`?

another reason i ask is that this part of the code is shared between the pgwire-level `Parse` logic and the sql-level PREPARE logic, so i just want to make sure it's ok to do this for both.
</blockquote></details>

Yeah good point. I moved the logic in `placeholders.Init()`, so that `PlaceholderTypesInfo.Types` are inited with `PlaceholderTypesInfo.TypeHints`, and let it be overridden by the deduced type later.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[`pkg/sql/conn_executor_prepare.go` line 270 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQp6SW8A0Yr-4WQD5_:-NAQp6SXCF8k8OD-cKup:bsf82c2) ([raw file](https://github.com/cockroachdb/cockroach/blob/790e765ee05abf1e082263ac1c3ee4fec4b7fe29/pkg/sql/conn_executor_prepare.go#L270)):*
> ```Go
> 		}
> 	}
> 	if err := p.semaCtx.Placeholders.Init(len(placeholderHints), placeholderHints); err != nil {
> ```

is this correct for the pgwire-prepare code path? i feel like it's possible that placeholderHints is nil in that path
___
*[`pkg/sql/sem/tree/placeholders.go` line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQoacV4j1gpa0QR3SQ:-NAQoacV4j1gpa0QR3SR:b-hue9hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c84c34cfccc16f6788672fe080c82a52456825b/pkg/sql/sem/tree/placeholders.go#L133)):*
> ```Go
> func (p *PlaceholderTypesInfo) SetType(idx PlaceholderIdx, typ *types.T) error {
> 	if t := p.Types[idx]; t != nil {
> 		if !typ.Equivalent(t) {
> ```

i don't see why we are changing this behavior. this seems like something we'd want to keep
___
*[`pkg/sql/sem/tree/placeholders.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQokAQ3cbtKmV1mhrC:-NAQokAQ3cbtKmV1mhrD:bfuhggv) ([raw file](https://github.com/cockroachdb/cockroach/blob/790e765ee05abf1e082263ac1c3ee4fec4b7fe29/pkg/sql/sem/tree/placeholders.go#L153)):*
> ```Go
> 		// deduced value afterwards.
> 		for i := 0; i < numPlaceholders; i++ {
> 			p.Types[i] = typeHints[i]
> ```

hm maybe i was wrong, since now i see a test failing with
```
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized:ERROR Code:XX000 Message:internal error: expected *DInt, found *tree.DString Detail:stack trace:
```


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
I changed the logic to extending the nil `Type` with `TypeHint`, but only when it's a sql-level `PREPARE` stmt (as opposed to a pgwire-level one). 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-:-NAeI9575B879em5lv14:b-zf4xru)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/conn_executor_prepare.go` line 270 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQp6SW8A0Yr-4WQD5_:-NARPtE-BK1-l4mJ02vo:b-913wtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/790e765ee05abf1e082263ac1c3ee4fec4b7fe29/pkg/sql/conn_executor_prepare.go#L270)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this correct for the pgwire-prepare code path? i feel like it's possible that placeholderHints is nil in that path
</blockquote></details>

Yeah, that makes sense. I now change it back to what we had, and only use `len(placeholderHints)` when the query is a sql `PREPARE` statement, where the hints will always be not-nil. 
___
*[`pkg/sql/plan_opt.go` line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAPyKya3Jdm7jAd2TY9:-NARV0yU6G7TGZghBm3e:b-yhzzp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/13e9135cb837ed67ab45a6bd6dbc1ed7ccf99067/pkg/sql/plan_opt.go#L191)):*
> this part of the code is shared between the pgwire-level Parse logic and the sql-level PREPARE logic

I now added a gate to restrict this logic only execute for the sql-level PREPARE logic.
___
*[`pkg/sql/logictest/testdata/logic_test/pg_catalog` line 4753 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAeHN5v3CF_sUTF5zwe:-NAeHN5v3CF_sUTF5zwf:b-1fria6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/logictest/testdata/logic_test/pg_catalog#L4753)):*
> ```
> 
> query TTTB
> SELECT name, statement, parameter_types, from_sql FROM pg_prepared_statements WHERE name LIKE 'args_%' ORDER BY 1,2
> ```

I added this test here because the `parameter_types` queried from `pg_prepared_statements` seem to be converted to oids when running in the Bazel Essential CI: 

```
        expected:
            args_deduce_type    PREPARE args_deduce_type (int, varchar, int, int) AS INSERT INTO t VALUES ($1, $2, $3)                   {bigint,"'character varying'",bigint,bigint}          true
            args_deduce_type_1  PREPARE args_deduce_type_1 (int, varchar, varchar) AS SELECT $1::INT8, $2::VARCHAR(10), $3::VARCHAR(20)  {bigint,"'character varying'","'character varying'"}  true
            args_test_few       PREPARE args_test_few (int, int) AS SELECT $1, $2::INT8                                                  {bigint,bigint}                                       true
            args_test_many      PREPARE args_test_many (int, int) AS SELECT $1                                                           {bigint,bigint}                                       true
        but found (query options: "") :
            args_deduce_type    PREPARE args_deduce_type (int, varchar, int, int) AS INSERT INTO t VALUES ($1, $2, $3)                   {20,1043,20,20}  true
            args_deduce_type_1  PREPARE args_deduce_type_1 (int, varchar, varchar) AS SELECT $1::INT8, $2::VARCHAR(10), $3::VARCHAR(20)  {20,1043,1043}   true
            args_test_few       PREPARE args_test_few (int, int) AS SELECT $1, $2::INT8                                                  {20,20}          true
            args_test_many      PREPARE args_test_many (int, int) AS SELECT $1                                                           {20,20}          true
```
Note that it didn't happen in Github CI and my local test run, but did in Bazel Essential CI, which is confusing.
___
*[`pkg/sql/sem/tree/placeholders.go` line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQoacV4j1gpa0QR3SQ:-NAeFQLv7UdMIrs2sMYp:b-dpovui) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c84c34cfccc16f6788672fe080c82a52456825b/pkg/sql/sem/tree/placeholders.go#L133)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't see why we are changing this behavior. this seems like something we'd want to keep
</blockquote></details>

I changed the logic back to "extending the Types with TypeHints" instead of `using TypeHints as the pre-set value for Type`, so I just changed this function back to what it was originally. 
___
*[`pkg/sql/sem/tree/placeholders.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAQokAQ3cbtKmV1mhrC:-NAeGIXV7hIW8vOJqckT:b-mt9mtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/790e765ee05abf1e082263ac1c3ee4fec4b7fe29/pkg/sql/sem/tree/placeholders.go#L153)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm maybe i was wrong, since now i see a test failing with
```
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized:ERROR Code:XX000 Message:internal error: expected *DInt, found *tree.DString Detail:stack trace:
```
</blockquote></details>

I think it failed because somehow we were just using the hint type (which is `string`) instead of the deduced type (int) from the original `generate_series` builtin func. The deduced type missed overriding the type hint. 


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 1 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @ZhouXing19)

___
*[`pkg/sql/sem/tree/placeholders.go` line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAep5og1fxVy3q19iSD:-NAep5og1fxVy3q19iSE:by7j8f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/sem/tree/placeholders.go#L163)):*
> ```Go
> 			return pgerror.Newf(
> 				pgcode.Syntax,
> 				"There should be at least one type hint for PREPARE statement",
> ```

`pgerrors` are user-facing errors, but this appears to be more of an internal error. Maybe use `errors.AssertionFailedf` instead.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/conn_executor_prepare.go` line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAfTRF23aKkcC8c9UJl:-NAfTRF23aKkcC8c9UJm:bcr5j7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/conn_executor_prepare.go#L75)):*
> ```Go
> 		parseCmd.RawTypeHints,
> 		PreparedStatementOriginWire,
> 		false /* fromSQL */)
> ```

the bool should not be needed. instead, the code can just check if the "origin" is `PreparedStatementOriginSQL`


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/sem/tree/placeholders.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAfTne75tYQz6NzoxTD:-NAfTne75tYQz6NzoxTE:bo5h48h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/sem/tree/placeholders.go#L101)):*
> ```Go
> 	// FromSQLPrepare is true when the placeholder is in a statement from a
> 	// PREPARE SQL stmt (rather than a pgwire prepare stmt).
> 	FromSQLPrepare bool
> ```

consider using the `PreparedStatementOrigin` type here


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/conn_executor_prepare.go` line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAfTRF23aKkcC8c9UJl:-NAfZHCb0pNTZyHB4oAN:b-dj0n0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/conn_executor_prepare.go#L75)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the bool should not be needed. instead, the code can just check if the "origin" is `PreparedStatementOriginSQL`
</blockquote></details>

Ah, i didn't notice we have that handy already. Will remove it.
___
*[`pkg/sql/sem/tree/placeholders.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAfTne75tYQz6NzoxTD:-NAfZLwn7qeqzUeujEKP:b41zarx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/sem/tree/placeholders.go#L101)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

consider using the `PreparedStatementOrigin` type here
</blockquote></details>

I don't think we can import `sql` pkg to `tree`, so should we keep using bool here?


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql: allow mismatch type numbers in PREPARE statement

Previously, we only allow having the same number of parameters and placeholders
in a PREPARE statement. This is not compatible with Postgres14's behavior.

This commit is to loosen the restriction and enable this compatibility.
We now take max(#placeholders, #parameters) as the true length
of parameters of the prepare statement. For each parameter, we first
have the type hint as the default value for each type. Then if the type can be
deduced from the query stmt, we let the deduced value override the type hint.

fixes #86375

Release justification: Low risk, high benefit changes to existing functionality
Release note (sql change): allow mismatch type numbers in PREPARE statement

rafiss reviewed (approved):
left an optional suggest. lgtm!

> Reviewed 1 of 15 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-:-NAivljTA-_UtQyphpKG:b3v8k48)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rafiss, and @ZhouXing19)

___
*[`pkg/sql/conn_executor_prepare.go` line 283 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAisjvU7z0IzBjXCUlp:-NAisjvV7PS_fTecsxsj:b-6ur6xn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d4cd8ed0d81d45e7902b1783a697f73ea00ddb9/pkg/sql/conn_executor_prepare.go#L283)):*
> ```Go
> 	if origin == PreparedStatementOriginSQL {
> 		p.semaCtx.Placeholders.FromSQLPrepare = true
> 		if p.semaCtx.Placeholders.TypeHints == nil {
> ```

does this idea work:

can we move this logic into `Init`? it might require a new signature like
 `func (p *PlaceholderInfo) Init(numPlaceholders int, typeHints PlaceholderTypes, fromSQLPrepare bool) error`

then there should be no need to have the assertion error, since `Init` is the function that's adding the `TypeHints` in the first place


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86904)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/conn_executor_prepare.go` line 283 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAisjvU7z0IzBjXCUlp:-NAjKb-fDk3E6wXG01Ej:b-g1hs46) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d4cd8ed0d81d45e7902b1783a697f73ea00ddb9/pkg/sql/conn_executor_prepare.go#L283)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

does this idea work:

can we move this logic into `Init`? it might require a new signature like
 `func (p *PlaceholderInfo) Init(numPlaceholders int, typeHints PlaceholderTypes, fromSQLPrepare bool) error`

then there should be no need to have the assertion error, since `Init` is the function that's adding the `TypeHints` in the first place

</blockquote></details>

Makes sense, done.
___
*[`pkg/sql/sem/tree/placeholders.go` line 163 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86904#-NAep5og1fxVy3q19iSD:-NAjKis-9QpYJc1DLtKC:bys5o2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07efd982ab1886686005df2b391a80570beeed6/pkg/sql/sem/tree/placeholders.go#L163)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`pgerrors` are user-facing errors, but this appears to be more of an internal error. Maybe use `errors.AssertionFailedf` instead.
</blockquote></details>

Yeah right, thanks for pointing this out! Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r=rafiss
